### PR TITLE
feat: add support for libc::memset

### PR DESCRIPTION
**Pull request by vishruth-thimmaiah**
closes: rust-lang/miri#4503

### CODE REVIEW

File: tests/pass-dep/libc/libc-mem.rs
Comment (RalfJung) [inline tests/pass-dep/libc/libc-mem.rs:1]: We'll also need a test in `fail-dep/libc/memset_null` doing a zero-sized memset on a null pointer -- which likely is UB for the C function, even though we allow it for the intrinsic.

File: src/shims/foreign_items.rs [hunk 1]
```diff
@@ -827,6 +827,22 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
...
+                let [ptr_dest, val, n] =
+                    this.check_shim_sig_lenient(abi, CanonAbi::C, link_name, args)?;
+                let ptr_dest = this.read_pointer(ptr_dest)?;
+                let val = this.read_scalar(val)?.to_i32()?;
+                let n = this.read_target_usize(n)?;
+                // The docs say val is "interpreted as unsigned char".
+                #[expect(clippy::as_conversions)]
+                let val = val as u8;
+
+                this.ptr_get_alloc_id(ptr_dest, 0)?;
```
Comment (RalfJung) [inline src/shims/foreign_items.rs:841]:
```suggestion
                // C requires that this must always be a valid pointer, even if `n` is zero, so we better check that.
                this.ptr_get_alloc_id(ptr_dest, 0)?;
```

**Review by RalfJung — COMMENTED** [review]
Thanks for the PR!

**Comment by RalfJung** [comment]
Looks good, thanks :)
